File: src/algorithm/scan.js
export function scan(requests, head, diskSize = 200, direction = 1) {
  let sequence = [head];
  let totalMovement = 0;

  // Step 1: Sort requests
  let sorted = [...requests].sort((a, b) => a - b);

  // Step 2: Find first request greater than head
  let index = 0;
  while (index < sorted.length && sorted[index] < head) {
    index++;
  }

  if (direction === 1) {
    // Move right
    for (let i = index; i < sorted.length; i++) {
      totalMovement += Math.abs(sorted[i] - head);
      head = sorted[i];
      sequence.push(head);
    }

    if (head !== diskSize - 1) {
      totalMovement += Math.abs((diskSize - 1) - head);
      head = diskSize - 1;
      sequence.push(head);
    }

    for (let i = index - 1; i >= 0; i--) {
      totalMovement += Math.abs(sorted[i] - head);
      head = sorted[i];
      sequence.push(head);
    }

  } else {
    // Move left
    for (let i = index - 1; i >= 0; i--) {
      totalMovement += Math.abs(sorted[i] - head);
      head = sorted[i];
      sequence.push(head);
    }

    if (head !== 0) {
      totalMovement += head;
      head = 0;
      sequence.push(head);
    }

    for (let i = index; i < sorted.length; i++) {
      totalMovement += Math.abs(sorted[i] - head);
      head = sorted[i];
      sequence.push(head);
    }
  }

  return {
    totalMovement,
    sequence
  };
}

File: backend/app/algorithms/scan.py
def scan(requests, head, disk_size, direction):
    sequence = [head]
    total_seek_time = 0
    current = head
    requests = sorted(requests)
    left = [request for request in requests if request < head]
    right = [request for request in requests if request >= head]
    if direction == "right":
        for request in right:
            total_seek_time += abs(request - current)
            current = request
            sequence.append(request)
        if current != disk_size - 1:
            total_seek_time += abs((disk_size - 1) - current)
            current = disk_size - 1
            sequence.append(current)
        for request in reversed(left):
            total_seek_time += abs(request - current)
            current = request
            sequence.append(request)
    else:
        for request in reversed(left):
            total_seek_time += abs(request - current)
            current = request
            sequence.append(request)
        if current != 0:
            total_seek_time += current
            current = 0
            sequence.append(current)
        for request in right:
            total_seek_time += abs(request - current)
            current = request
            sequence.append(request)
    average_seek_time = total_seek_time / len(requests)
    return {
        "sequence": sequence,
        "total_seek_time": total_seek_time,
        "average_seek_time": average_seek_time
    }